### ISSUE TITLE
There should be a caching wrapfs

### ISSUE THREAD

**Issue by hmeine** [PyFilesystem/pyfilesystem#223]
For networking-based filesystems, one is often interested in some kind of local cache.  As far as I can see, that would be another good use case for `fs.wrapfs`, right?

**Comment by travcunn** [comment]
Yes `fs.wrapfs` would be perfect. The implementation really depends on what type of filesystem you're trying to wrap and your use case though.

**Comment by arekbulski** [comment]
There are classes for caching single files and filesystem metadata, but not complete filesystem:
https://github.com/PyFilesystem/pyfilesystem/blob/master/fs/remote.py

**Comment by hmeine** [comment]
Yes, I think `CacheFS` comes close, but I would like to have an additional cache of files that were read.  So, one would configure a local cache directory and maximum cache size (after which the oldest files are purged), and all files would first be copied there and reused instead of re-downloaded.  The metadata would obviously be solved by `CacheFS[Mixin]` already.

**Comment by mguillau** [comment]
Hi,

I'm working on a WrapFS subclass implementing the LRU caching scheme [(wikipedia entry)](https://en.wikipedia.org/wiki/Cache_algorithms#Examples). On top of the wrapped FS, one provides another FS used for caching and the max size of the cache. The consistency is checked based on md5, so it's compatible with s3's etag. Would you consider a pull request when I'm done?

M.

**Comment by arekbulski** [comment]
I think you could close this issue, unless there you have some code to submit. You made a point.

**Comment by lurch** [comment]
IMHO this is a perfectly valid suggestion / request, so there's no point closing it.

**Comment by arekbulski** [comment]
I agree. Some maintainers close less prospective issues when there is no active work to discuss.
